runtime/description.js

```diff
@@ -172,23 +180,27 @@ export class DescriptionFormatter {
   _joinDescriptions(strings) {
     let nonEmptyStrings = strings.filter(str => str);
     let count = nonEmptyStrings.length;
-    // Combine descriptions into a sentence:
-    // "A."
-    // "A and b."
-    // "A, b, ..., and z." (Oxford comma ftw)
-    let delim = ['', '', ' and ', ', and '][Math.min(3, count)];
-    const lastString = nonEmptyStrings.pop();
-    return `${nonEmptyStrings.join(', ')}${delim}${lastString}`;
+    if (count > 0) {
+      // Combine descriptions into a sentence:
+      // "A."
+      // "A and b."
+      // "A, b, ..., and z." (Oxford comma ftw)
+      let delim = ['', '', ' and ', ', and '][Math.min(3, count)];
+      const lastString = nonEmptyStrings.pop();
+      return `${nonEmptyStrings.join(', ')}${delim}${lastString}`;
+    }
   }
 
   _joinTokens(tokens) {
     return tokens.join('');
   }
 
   _capitalizeAndPunctuate(sentence) {
-    // "Capitalize, punctuate." (if the sentence doesn't end with a punctuation character).
-    let last = sentence.length - 1;
-    return `${sentence[0].toUpperCase()}${sentence.slice(1, last)}${sentence[last]}${sentence[last].match(/[a-z0-9\(\)'>\]]/i) ? '.' : ''}`;
+    if (sentence) {
```

Comment:
If !sentence maybe we should still return something, or assert?

Reply:
changed if to assert here and moved the if check to the caller method. (it is ok, i think, for it to be undefined in the caller method, because it's a generic one, and the fallback to recipe name or default string is at a higher level)